finetune/input_pipeline.py

```diff
@@ -0,0 +1,209 @@
+
+    @abstractmethod
+    def _target_encoder(self):
+        # Overridden by subclass to produce the right target encoding for a given target model.
+        raise NotImplementedError
+
+    def feed_shape_type_def(self):
+        TS = tf.TensorShape
+        return ({"tokens": tf.int32, "mask": tf.float32}, tf.float32), (
+            {"tokens": TS([self.config.max_length, 2]), "mask": TS([self.config.max_length])}, TS([self.target_dim]))
+
+    def _array_format(self, encoded_output, pad_token=PAD_TOKEN):
+        """
+        Returns numpy array of token idxs and corresponding mask
+        Returned `x` array contains two channels:
+            0: byte-pair encoding embedding
+            1: positional embedding
+        """
+        seq_length = len(encoded_output.token_ids)
+        x = np.zeros((self.config.max_length, 2), dtype=np.int32)
+        mask = np.zeros((self.config.max_length), dtype=np.float32)
+
+        if encoded_output.labels is not None:
+            labels_arr = np.empty((self.config.max_length), dtype='object')
+            labels_arr.fill(pad_token)
+        else:
+            labels_arr = None
+
+        # BPE embedding
+        x[:seq_length, 0] = encoded_output.token_ids
+        # masking: value of 1 means "consider this in cross-entropy LM loss"
+        mask[1:seq_length] = 1
+        if encoded_output.labels:
+            labels_arr[:seq_length] = encoded_output.labels
+        # positional_embeddings
+        x[:, 1] = np.arange(self.encoder.vocab_size, self.encoder.vocab_size + self.config.max_length)
+
+        return ArrayEncodedOutput(
+            token_ids=x,
+            tokens=encoded_output.tokens,
+            labels=labels_arr,
+            char_locs=encoded_output.char_locs,
+            mask=mask,
+        )
+
+    def _adapt_feats_for_distributed(self, feats_labels):
+        if type(feats_labels) == tuple:
+            feats, labels = feats_labels
+            feats["labels"] = labels
+            return feats
+
+    def text_to_tokens_mask(self, X, Y=None):
+        out_gen = self._text_to_ids(X)
+        for out in out_gen:
+            feats = {"tokens": out.token_ids, "mask": out.mask}
+            if Y is None:
+                yield feats
+            else:
+                yield feats, self.label_encoder.transform([Y])[0]
+
+    def _post_data_initialization(self, Y):
+        self.label_encoder = self._target_encoder()
+        if not callable(Y):
+            Y_fit = Y
+            self.config.dataset_size = len(Y)
+            self.label_encoder.fit_transform(Y)
+        else:
+            try:
+                self.config.dataset_size = len(Y())
+            except TypeError:
+                logging.warning(
+                    "Generator input function does not have a length, falling back to default in config of {}".format(
+                        self.config.dataset_size))
+            Y_fit = list(itertools.islice(Y(), 100))  # TODO find a more principled way to do this?
```

Comment:
This is absolutely going to cause problems on datasets w/ rare classes.  One of the few issues so far that I think requires a fix before merge.

Reply:
Is it possible to ask for a list of possible target values if the input is a callable?

Reply:
Or... just up this to a value that's > 100.

Reply:
Resolution -- make this number hella big or consume all targets.